projects/driver_controls/src/driver_io.c

```diff
@@ -0,0 +1,83 @@
+#include <stddef.h>
+
+#include "adc.h"
+#include "gpio_it.h"
+#include "driver_io.h"
+
+// Arbitrary thresholds for gas pedal
+#define COAST_THRESHOLD 1000
+#define DRIVE_THRESHOLD 3000
+
+typedef struct DriverIOSettings {
+  GPIOAddress address;
+  GPIODir direction;
+  GPIOAltFn alt_function;
+  InterruptEdge edge;
+  DriverIOCallback callback;
+  DriverIOData data;
+} DriverIOSettings;
+
+static void prv_init_pin(DriverIOSettings *settings) {
+  GPIOSettings gpio_settings = { settings->direction, GPIO_STATE_LOW,
+                                  GPIO_RES_NONE, settings->alt_function };
+  InterruptSettings it_settings = { INTERRUPT_TYPE_INTERRUPT, INTERRUPT_PRIORITY_LOW };
+
+  gpio_init_pin(&settings->address, &gpio_settings);
+
+  if (settings->callback != NULL) {
+    gpio_it_register_interrupt(&settings->address, &it_settings, settings->edge,
+                                settings->callback, &settings->data);
```

Comment:
Right now, `settings->data` is stack-allocated. When it runs the callback and you try to edit the context, it'll refer to undefined memory.